airbyte-ci/connectors/live-tests/src/live_tests/regression_tests/conftest.py

```diff
@@ -0,0 +1,579 @@
+
+
+@pytest.fixture(scope="session")
+def connector_config(request: SubRequest) -> Optional[SecretDict]:
+    return get_connector_config(get_option_or_fail(request, "--config-path"))
+
+
+@pytest.fixture(scope="session")
+def state(request: SubRequest) -> Optional[dict]:
+    return get_state(get_option_or_fail(request, "--state-path"))
+
+
+@pytest.fixture(scope="session")
+def dagger_connection(request: SubRequest) -> dagger.Connection:
+    return dagger.Connection(dagger.Config(log_output=request.config.stash[DAGGER_LOG_PATH].open("w")))
+
+
+@pytest.fixture(scope="session")
+async def dagger_client(dagger_connection: dagger.Connection) -> AsyncIterable[dagger.Client]:
+    async with dagger_connection as client:
+        yield client
+
+
+@pytest.fixture(scope="session")
+async def control_connector(dagger_client: dagger.Client, connector_image: str, control_version: str) -> ConnectorUnderTest:
+    return await get_connector_under_test(dagger_client, f"{connector_image}:{control_version}")
+
+
+@pytest.fixture(scope="session")
+async def target_connector(dagger_client, connector_image, target_version) -> ConnectorUnderTest:
+    return await get_connector_under_test(dagger_client, f"{connector_image}:{target_version}")
+
+
+@pytest.fixture
+def spec_control_execution_inputs(
+    control_connector: ConnectorUnderTest,
+) -> ExecutionInputs:
+    return ExecutionInputs(connector_under_test=control_connector, command=Command.SPEC)
+
+
+@pytest.fixture
+def spec_control_connector_runner(dagger_client: dagger.Client, spec_control_execution_inputs: ExecutionInputs) -> ConnectorRunner:
+    return ConnectorRunner(dagger_client, **spec_control_execution_inputs.to_dict())
+
+
+@pytest.fixture
+async def spec_control_execution_result(
+    record_property: Callable,
+    request: SubRequest,
+    test_artifacts_directory: Path,
+    spec_control_execution_inputs: ExecutionInputs,
+    spec_control_connector_runner: ConnectorRunner,
+    session_start_timestamp: int,
+) -> ExecutionResult:
+    logging.info(f"Running spec for control connector {spec_control_execution_inputs.connector_under_test.name}")
+    execution_result = await spec_control_connector_runner.run()
+    execution_report = await persist_report(
+        request,
+        test_artifacts_directory,
+        spec_control_execution_inputs,
+        execution_result,
+        session_start_timestamp,
+    )
+    record_property("control_execution_report", execution_report)
+    return execution_result
+
+
+@pytest.fixture
+def spec_target_execution_inputs(
+    target_connector: ConnectorUnderTest,
+) -> ExecutionInputs:
+    return ExecutionInputs(connector_under_test=target_connector, command=Command.SPEC)
+
+
+@pytest.fixture
+def spec_target_connector_runner(dagger_client: dagger.Client, spec_target_execution_inputs: ExecutionInputs) -> ConnectorRunner:
+    return ConnectorRunner(dagger_client, **spec_target_execution_inputs.to_dict())
+
+
+@pytest.fixture
```

Comment:
Declaring this (and execution results for other commands) as a pytest fixture has the effect of executing them before any tests run. Is there a reason you prefer to do it this way rather than getting the result as part of the test? I have a slight preference for the latter - to me, executing the calls feels like part of the test behavior, so should fail during execution of a specific test.

Also, will declaring these as fixtures end up running all commands, even if we only want to run a single test? Or will they be executed lazily?

Reply:
I think its indeed a fair point of discussion.
As these fixtures are indeed lazily used on test request I found it useful to declare this logic in fixture.
IMO it avoids a lot of code duplication if we create multiple tests performing assertions on command execution results, and it makes sure the execution report is created.